### PR TITLE
AgentConfigManager implementation

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

AgentConfigManager implementation

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/configurations/schemas/definitions.json [hunk 1]
```diff
@@ -150,6 +153,21 @@
...
+              "type": ["boolean", "null"]
+            },
+            "version": {
+              "type": "number"
+            },
+            "handlers": {"type": "object"},
+            "formatters": {"type": "object"},
+            "loggers": {"type": "object"}
+        }
     }
```
Comment (DavidMinarsch) [inline aea/configurations/schemas/definitions.json:171]: I don't know if this is the full schema...Have you confirmed that it is?

Comment (DavidMinarsch) [reply aea/configurations/schemas/definitions.json:171]: https://docs.python.org/3/library/logging.html

File: aea/configurations/base.py [hunk 2]
```diff
@@ -327,6 +342,75 @@ def _create_or_update_from_json(
...
+                errors.append(
+                    f"For {'.'.join(path)} `{type(current_value)}` data type is expected, but `{type(new_value)}` was provided!"
+                )
+        return errors
+
+    def make_resulting_config_data(self, overrides: Dict) -> Dict:
+        """Make config data with overrides applied."""
+        current_config = self.json
+        recursive_update(current_config, overrides, allow_new_values=True)
+        return current_config
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:391]: Mention in docstring that this method does *not* update the object!

File: aea/configurations/base.py [hunk 3]
```diff
@@ -597,10 +683,16 @@ def _create_or_update_from_json(
...
+
         if instance is None:
             instance = cls(**params)  # type: ignore
         else:
             instance.__init__(**params)  # type: ignore
+
+        if not instance.directory:
+            instance.directory = directory
+
         return instance
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:696]: extract in class method?

Comment (DavidMinarsch) [reply aea/configurations/base.py:696]: allows for reuse below and separation

File: aea/configurations/base.py [hunk 4]
```diff
@@ -925,7 +1031,14 @@ def _create_or_update_from_json(
 
         return instance
 
-    def update(self, data: Dict) -> None:
+    def make_resulting_config_data(self, overrides: Dict) -> Dict:
+        """Make config data with overrides applied."""
+        self._update(overrides)
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:1036]: Ok, but here we actually do change the object...

File: aea/configurations/base.py [hunk 5]
```diff
@@ -965,13 +1081,34 @@ def _update_skill_component_config(type_plural: str, data: Dict):
...
         _update_skill_component_config("behaviours", data)
         _update_skill_component_config("handlers", data)
         _update_skill_component_config("models", data)
         self.is_abstract = data.get("is_abstract", self.is_abstract)
 
+    def get_overridable(self) -> dict:
+        """Get overrideable confg data."""
+        result = {}
+        current_config_data = self.json
+        if "is_abstract" in current_config_data:
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:1099]: use constant

File: aea/configurations/base.py [hunk 6]
```diff
@@ -1001,6 +1138,10 @@ class AgentConfig(PackageConfiguration):
             "storage_uri",
         ]
     )
+    CHECK_EXCLUDES = [
+        ("private_key_paths",),
+        ("default_routing",),
+    ]
```
Comment (DavidMinarsch) [inline aea/configurations/base.py:1144]: Why don't we check?

File: aea/cli/utils/constants.py [hunk 7]
```diff
@@ -80,6 +79,6 @@
     CONNECTIONS: DEFAULT_CONNECTION_CONFIG_FILE,
     CONTRACTS: DEFAULT_CONTRACT_CONFIG_FILE,
 }  # type: Dict[str, str]
-FALSE_EQUIVALENTS = ["f", "false", "False"]
+FALSE_EQUIVALENTS = ["f", "false", "False", "0"]
```
Comment (DavidMinarsch) [inline aea/cli/utils/constants.py:82]: nice!

**Review by DavidMinarsch — COMMENTED** [review]
LGTM, some comments to address and discuss. Thanks!

File: aea/configurations/base.py (hunk 6 above)
Comment (solarw) [reply aea/configurations/base.py:1144]: cause can contain attributes not mentioned in the current config

**Pull request merged**